### PR TITLE
Migrate flow to `Reconcilable` interface and generic adapter reconciler

### PR DESCRIPTION
**Pull request by antoineco**
Same as #682 but for types in the `flow` API.

Part two of the grand plan mentioned in #678.

With `flow` components, I had to get more granular in the way I handle components that may or may not define a sink. This shows in `pkg/apis/flow/v1alpha1/interfaces.go` and in `pkg/flow/reconciler/common/reconcile.go` with the handling of the various sub-interfaces.
The result is a version of the generic adapter reconciler that can handle both sources _and_ targets, which means I'm very close to merging all these in my next PR and delete a lot of redundant code (which was the main goal all along).

Tasks:

- [x] Copy `reconciler/common` to `pkg/flow` and perform the required adjustments
- [x] Implement the `Reconcilable` interface in all target types
- [x] Migrate target reconcilers
  - [x] JQTransformation
  - [x] Synchronizer
  - [x] Transformation
  - [x] XMLToJSONTransformation
  - [x] XSLTTransformation

### CODE REVIEW

File: pkg/flow/reconciler/common/semantic/semantic.go [hunk 1]
```diff
@@ -36,12 +36,12 @@ import (
 // DeepDerivative comparisons to work as expected.
```
Comment (cab105) [inline pkg/flow/reconciler/common/semantic/semantic.go:36]: Fix copyright

File: pkg/flow/reconciler/common/semantic/semantic_test.go [hunk 2]
```diff
@@ -173,47 +174,49 @@ func TestServiceEqual(t *testing.T) {
 	}
 
 	for name, tc := range testCases {
-		//nolint:scopelint
+		// //nolint:scopelint
```
Comment (cab105) [inline pkg/flow/reconciler/common/semantic/semantic_test.go:177]: Don't think we need the second `//` chunk

**Review by cab105 — COMMENTED** [review]
Left a few minor nits that I saw during the review. This does make writing new adapters sooooo much easier.

Will caveat that I only commented on a couple of the copyright nits that I observed, but you will definitely want to double check the files to ensure it is for 2022 instead of 2021.

File: pkg/apis/flow/interfaces.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
+/*
+Copyright 2021 TriggerMesh Inc.
```
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/apis/flow/interfaces.go]: I relocated all interfaces to `v1alpha1/interfaces.go` (which makes more sense), which has now the correct copyright.
Marking as resolved.

File: pkg/apis/flow/interfaces.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flow
+
+// IntegrationTarget is implemented by types which are integration targets.
```
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/apis/flow/interfaces.go]: This interface name didn't translate very well to components other than targets. I renamed it to `EventReceiver`.
In fact, the right thing to do would be to use that name in targets too. That will resolve itself in the next PR, once I move those interfaces into a package that's common to all TriggerMesh components.

File: pkg/apis/flow/v1alpha1/common_types.go [hunk 5]
```diff
@@ -18,20 +18,55 @@ package v1alpha1
...
+
 // ValueFromField is a struct field that can have its value either defined
 // explicitly or sourced from another entity.
 type ValueFromField struct {
 	// Optional: no more than one of the following may be specified.
 
 	// Field value.
 	// +optional
-	Value *string `json:"value,omitempty"`
+	Value string `json:"value,omitempty"`
```
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/apis/flow/v1alpha1/common_types.go:35]: The rationale is the same as in Kubernetes' `corev1.EnvVar` (which was my inspiration for creating this type originally): it is enough to check for `""` because an empty string will be omitted anyway while (de-)serializing to/from JSON.

With that being said, the removal of the pointer type here is only a result of me copying the type from `sources` / `routing` / `target` for consistency. I'm totally fine reverting to a pointer if we agree that this becomes a pointer in other component types too.

File: pkg/apis/flow/v1alpha1/common_test.go [hunk 6]
```diff
@@ -16,9 +16,7 @@ limitations under the License.
```
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/apis/flow/v1alpha1/common_test.go:16]: If the agreement is that we should update the copyright year on every file we _modify_ and in _every PR_, shall we simply do a huge string replacement over the repo and be done with it?

[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/apis/flow/v1alpha1/common_test.go:16]: Done in 3e4c4aa0 (for the files touched in this PR only)

File: pkg/flow/reconciler/common/adapter.go [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/system"
+	"knative.dev/serving/pkg/apis/serving"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+
+	"github.com/triggermesh/triggermesh/pkg/apis/flow/v1alpha1"
+	"github.com/triggermesh/triggermesh/pkg/flow/reconciler/common/resource"
+)
+
+const metricsPrometheusPortKsvc uint16 = 9092
```
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/flow/reconciler/common/adapter.go:43]: Done in 1e6d609e

File: pkg/flow/reconciler/common/base.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		if !ok {
+			return false
+		}
+
+		ls := CommonObjectLabels(typ).AsSelectorPreValidated()
+		return ls.Matches(labels.Set(object.GetLabels()))
+	}
+}
+
+// isInNamespace returns a function that filters all objects which are not in
```
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/flow/reconciler/common/base.go:144]: Do you mean that the comment should read "filter out" instead of "filter"?

In plain English, this is what this function is used for:
For every object found in the informer's cache, execute `isInNamespace(obj)` and keep only the ones for which the function returned true (meaning the object _is_ in the desired namespace).

I could also rephrase to
> isInNamespace returns a filter function which returns true if the object is in the given namespace.

File: pkg/flow/reconciler/common/semantic/semantic.go (hunk 1 above)
Comment (antoineco) [reply pkg/flow/reconciler/common/semantic/semantic.go:36]: Done in 3e4c4aa0

File: pkg/flow/reconciler/common/resource/object_test.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/ptr"
+)
+
+const (
+	tNs   = "testns"
+	tName = "testname"
```
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/flow/reconciler/common/resource/object_test.go]: That makes sense 👍 
Done in 4462a51f

File: pkg/flow/reconciler/common/semantic/semantic_test.go (hunk 2 above)
Comment (antoineco) [reply pkg/flow/reconciler/common/semantic/semantic_test.go]: Nice catch!
Done in e1dea361

File: pkg/apis/flow/v1alpha1/common_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cab105) [reply pkg/apis/flow/v1alpha1/common_types.go:35]: I'm totally game with keeping this as a generic string and do get your point on the serialization/deserializtion. Upon first glance, I thought this was introducing an inconsistency instead of correcting for it.

File: pkg/flow/reconciler/common/base.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (cab105) [reply pkg/flow/reconciler/common/base.go:144]: > Do you mean that the comment should read "filter out" instead of "filter"?

Correct. As the comment is written, I understood it to be that you were returning a function that would return object _not_ in the namespace.

Your rephrased comment reads much better.

**Review by cab105 — APPROVED** [review]
Preapproving with my comments addressed.

File: pkg/flow/reconciler/common/base.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/flow/reconciler/common/base.go]: Done in 45306591
